### PR TITLE
Make HttpClient respect the path in the client URI (#375)

### PR DESCRIPTION
**Pull request by lespinsideg**
issue number : #375 
- concatenate paths in client URI and HttpClient request.
- add test cases for modification.

### CODE REVIEW

File: core/src/test/java/com/linecorp/armeria/client/http/HttpClientDelegateTest.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import com.linecorp.armeria.common.http.HttpHeaders;
+import com.linecorp.armeria.common.http.HttpMethod;
+import com.linecorp.armeria.common.http.HttpRequest;
+
+import io.netty.channel.DefaultEventLoop;
+
+public class HttpClientDelegateTest {
+    @Test
+    public void givenDifferentPathInContextAndHttpRequest_whenExecute_thenSetsPathInContextToRequestHeader()
+            throws Exception {
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/client/http/HttpClientDelegateTest.java]: Could you explain why this test case is required? 1) When does this case happen? 2) Is this test case related with respecting the client URI? If so, how?

**Comment by lespinsideg** [comment]
@trustin 
I wrote that test case because of the logic I added to DefaultHttpClient(after passing concatenated URI to execute method in UserClient, request paths in HttpHeader and ClientRequestContext does not synchronized. But after reviewing the code, I found that it is better to sync the paths in HttpHeader and ClientRequestContext on DefaultHttpClient. I reverted the commits and modified the code. Please review that.

File: core/src/main/java/com/linecorp/armeria/client/http/DefaultHttpClient.java [hunk 2]
```diff
@@ -46,7 +46,10 @@ public HttpResponse execute(HttpRequest req) {
     }
 
     private HttpResponse execute(EventLoop eventLoop, HttpRequest req) {
-        return execute(eventLoop, req.method().name(), req.path(), "", req, cause -> {
+        final String respectingClientUriPath = uri().isOpaque() ? req.path() : uri().getPath() + req.path();
+        req.path(respectingClientUriPath);
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/http/DefaultHttpClient.java]: - Could we avoid resulting in the path that contains `//`? i.e. check `uri().getPath()` does not end with `/` and `req.path()` starts with `/`, etc
- `respectingClientUriPath` -> How about just `path`?

Comment (lespinsideg) [inline core/src/main/java/com/linecorp/armeria/client/http/DefaultHttpClient.java]: @trustin Added remove consecutive slashes logic and test case.

File: core/src/main/java/com/linecorp/armeria/client/http/DefaultHttpClient.java [hunk 3]
```diff
@@ -78,4 +85,27 @@ HttpResponse execute(EventLoop eventLoop, AggregatedHttpMessage aggregatedReq) {
         req.close();
         return execute(eventLoop, req);
     }
+
+    private static String concatPath(String path1, String path2) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/http/DefaultHttpClient.java]: This method is too heavyweight. What I would expect is just a simple stuff like this:

```java
private static String concatPath(String path1, String path2) {
  path1 = path1 == null ? "" : path1;
  path2 = path2 == null ? "" : path2;

  final String concatPath;
  if (path1.isEmpty()) {
    if (path2.isEmpty()) {
        concatPath = "/";
    } else if (path2.charAt(0) == '/') {
        concatPath = path2;
    } else {
        concatPath = '/' + path2;
    }
  } else if (path2.isEmpty()) {
    concatPath = path1;
  } else if (path1.charAt(path1.length() - 1) == '/') {
    if (path2.charAt(0) == '/') {
      concatPath = new StringBuilder(path1.length() + path2.length() - 1).append(path1).append(path2, 1, path2.length()).toString();
    } else {
      concatPath = path1 + path2;
    }
  } else {
    if (path2.charAt(0) == '/') {
      concatPath = path1 + path2;
    } else {
      concatPath = new StringBuilder(path1.length() + path2.length() +1).append(path1).append('/').append(path2).toString();
    }
  }
}
```

**Comment by trustin** [comment]
@lespinsideg Meanwhile, could you sign the ICLA, if you didn't yet? http://feedback-admin-internal.linecorp.com/support/feedback/enquete/

**Comment by trustin** [comment]
Thanks, @lespinsideg !

**Comment by trustin** [comment]
@lespinsideg Here's the follow-up commit: https://github.com/line/armeria/commit/4c214cdb597af4068fcb270aa36404c9a9e96548

**Comment by lespinsideg** [comment]
Thank you 👍 @trustin

**Comment by trustin** [comment]
> Thank you 👍 @trustin 

My pleasure, @lespinsideg. Looking forward to your continued involvement with Armeria, regardless whether it's a pull request or a feature request!
